README.md
# DbdOnto

Ontologies for [Dbd].

First a limited context ontology is constructed, because it is required for all facts. This context ontology also has a "meta" context itself (using the context predicates defined in itself, to bootstrap the system).

The context ontology uses one meta:defines_predicate attribute from a new Meta ontology. This is needed to describe the definition of predicates, since predicates cannot be the subject of Fact (in Dbd, a subject is an abstract UUID, not a URI, so a link is needed between a resource (all facts with the same subject) and the predicate (that is then the object for this fact that defines the predicate)).

[![Gem Version](https://badge.fury.io/rb/dbd_onto.png)](http://badge.fury.io/rb/dbd_onto)
[![Build Status](https://travis-ci.org/petervandenabeele/dbd_onto.png?branch=master)](http://travis-ci.org/petervandenabeele/dbd_onto)

## Installation

Add this line to your application's Gemfile:

    gem 'dbd_onto'

## Usage

E.g. use the predicates defined in DbdOnto::Context to fully populate a Context.

## Examples

The ontology is stored as a [Dbd]::Graph. Displaying it in short format:

```
$ irb
2.0.0p247 :001 > require 'dbd_onto'
 => true
2.0.0p247 :002 > DbdOnto::Meta.new.each {|f| puts f.short} ; nil
 cont ] : 36c0b50f : context:visibility       : public
[ cont ] : 36c0b50f : context:encryption       : clear
[ cont ] : 36c0b50f : context:license          : Copyright 2013 Peter Vandenabeele (http://vandenabeele.com), licensed under MIT
[ cont ] : 36c0b50f : dc:source                : https://github.com/petervandenabeele/dbd/blob/d37360070e7f8e61a19c2bca210c881a15
[ cont ] : 36c0b50f : dc:creator               : Peter Vandenabeele (@peter_v)
[ cont ] : 36c0b50f : dcterms:created          : 2013-08-09 21:45:00 UTC
[ cont ] : 36c0b50f : rdfs:comment             : This meta context covers a creation period starting Aug 2013 until ...
36c0b50f : 619e69d4 : meta:defines_predicate   : meta:defines_predicate
36c0b50f : 619e69d4 : rdfs:label               : Defines predicate
36c0b50f : 619e69d4 : dbd:rdf_uri              : https://github.com/petervandenabeele/dbd_onto/blob/master/data/meta.md
36c0b50f : 619e69d4 : rdfs:comment             : Needed to bootstrap ontology definitions
36c0b50f : 619e69d4 : meta:defines_predicate   : meta:predicate_used
36c0b50f : 619e69d4 : rdfs:label               : Predicate used
36c0b50f : 619e69d4 : dbd:rdf_uri              : https://github.com/petervandenabeele/dbd_onto/blob/master/data/meta.md
36c0b50f : 619e69d4 : rdfs:comment             : Is this predicate used in the UI for creating new resources?
 => nil
2.0.0p247 :005 >   DbdOnto::Context.new.each {|f| puts f.short} ; nil
[ cont ] : 36c0b50f : context:visibility       : public
[ cont ] : 36c0b50f : context:encryption       : clear
[ cont ] : 36c0b50f : context:license          : Copyright 2013 Peter Vandenabeele (http://vandenabeele.com), licensed under MIT
[ cont ] : 36c0b50f : dc:source                : https://github.com/petervandenabeele/dbd/blob/d37360070e7f8e61a19c2bca210c881a15
[ cont ] : 36c0b50f : dc:creator               : Peter Vandenabeele (@peter_v)
[ cont ] : 36c0b50f : dcterms:created          : 2013-08-09 21:45:00 UTC
[ cont ] : 36c0b50f : rdfs:comment             : This meta context covers a creation period starting Aug 2013 until ...
36c0b50f : 03e72b0c : meta:defines_predicate   : context:visibility
36c0b50f : 03e72b0c : rdfs:label               : Visibility
36c0b50f : 03e72b0c : dbd:rdf_uri              : https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md
36c0b50f : 03e72b0c : rdfs:comment             : E.g. public, private, professional, ...
36c0b50f : 735ae2a2 : meta:defines_predicate   : context:encryption
36c0b50f : 735ae2a2 : rdfs:label               : Encryption
36c0b50f : 735ae2a2 : dbd:rdf_uri              : https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md
36c0b50f : 735ae2a2 : rdfs:comment             : E.g. clear, encrypted, AES-256, my_very_secret_key
36c0b50f : 5b2c73e5 : meta:defines_predicate   : context:license
36c0b50f : 5b2c73e5 : rdfs:label               : License
36c0b50f : 5b2c73e5 : dbd:rdf_uri              : https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md
36c0b50f : 5b2c73e5 : rdfs:comment             : E.g. Copyright 2013 Peter Vandenabeele (http://vandenabeele.com), licensed under
36c0b50f : ffc41d9c : meta:defines_predicate   : dc:source
36c0b50f : ffc41d9c : rdfs:label               : Source
36c0b50f : ffc41d9c : dbd:rdf_uri              : https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md
36c0b50f : ffc41d9c : rdfs:comment             : E.g. Entered manually by @peter_v (also see Dublin Core)
36c0b50f : 2b5008b0 : meta:defines_predicate   : dc:creator
36c0b50f : 2b5008b0 : rdfs:label               : Creator
36c0b50f : 2b5008b0 : dbd:rdf_uri              : https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md
36c0b50f : 2b5008b0 : rdfs:comment             : E.g. @peter_v (also see Dublin Core)
36c0b50f : 4aba46de : meta:defines_predicate   : dcterms:created
36c0b50f : 4aba46de : rdfs:label               : Created
36c0b50f : 4aba46de : dbd:rdf_uri              : https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md
36c0b50f : 4aba46de : rdfs:comment             : E.g. 2013-10-20 (also see Dublin Core Terms)
 => nil
2.0.0p247 :007 >   DbdOnto::Dbd.new.each {|f| puts f.short} ; nil
[ cont ] : 36c0b50f : context:visibility       : public
[ cont ] : 36c0b50f : context:encryption       : clear
[ cont ] : 36c0b50f : context:license          : Copyright 2013 Peter Vandenabeele (http://vandenabeele.com), licensed under MIT
[ cont ] : 36c0b50f : dc:source                : https://github.com/petervandenabeele/dbd/blob/d37360070e7f8e61a19c2bca210c881a15
[ cont ] : 36c0b50f : dc:creator               : Peter Vandenabeele (@peter_v)
[ cont ] : 36c0b50f : dcterms:created          : 2013-08-09 21:45:00 UTC
[ cont ] : 36c0b50f : rdfs:comment             : This meta context covers a creation period starting Aug 2013 until ...
36c0b50f : 619e69d4 : meta:defines_predicate   : dbd:rdf_uri
36c0b50f : 619e69d4 : rdfs:label               : RDF URI
36c0b50f : 619e69d4 : dbd:rdf_uri              : https://github.com/petervandenabeele/dbd_onto/blob/master/data/dbd.md
36c0b50f : 619e69d4 : rdfs:comment             : points to an RDF URI type, without the <> characters
 => nil
2.0.0p247 :011 > DbdOnto::Schema.new.take(14).each {|f| puts f.short} ; nil
[ cont ] : 37c0b50f : context:visibility       : public
[ cont ] : 37c0b50f : context:encryption       : clear
[ cont ] : 37c0b50f : context:license          : Copyright the sponsors of Schema.org: Google, Inc., Yahoo, Inc., and Microsoft C
[ cont ] : 37c0b50f : dc:source                : http://schema.org
[ cont ] : 37c0b50f : dc:creator               : Peter Vandenabeele (@peter_v)
[ cont ] : 37c0b50f : dcterms:created          : 2013-10-15 21:56:00 UTC
37c0b50f : d00871dc : meta:defines_predicate   : schema:about
37c0b50f : d00871dc : dbd:rdf_uri              : http://schema.org/about
37c0b50f : d00871dc : rdfs:label               : about
37c0b50f : d00871dc : rdfs:comment             : The subject matter of the content.
37c0b50f : fb6b6a77 : meta:defines_predicate   : schema:acceptedPaymentMethod
37c0b50f : fb6b6a77 : dbd:rdf_uri              : http://schema.org/acceptedPaymentMethod
37c0b50f : fb6b6a77 : rdfs:label               : acceptedPaymentMethod
37c0b50f : fb6b6a77 : rdfs:comment             : The payment method(s) accepted by seller for this offer.
 => nil
```

[Dbd]:              https://github.com/petervandenabeele/dbd#readme


HISTORY.txt
0.0.1 (2013-06-03)
=====

* A basic DbdOnto::Provenance ontology
* bootstrapping with a self referential provenance ontology

0.0.2 (2013-08-09)
=====

* Update to Dbd 0.0.14 (Provenance => Context)
* use dc:creator (instead of dcterms:creator that has a specific range)

0.0.3 (2013-09-03)
=====

* DbdOnto::Meta now has static data (constant id, subject, ...)

0.0.4 (2013-09-94)
=====

* DbdOnto::Meta and Context are now fully static data
* simplification of code (using Dbd 0.0.19)

0.0.5 (2013-10-14)
=====

* Refactor (remove an unused method)

0.0.6 (2013-10-20)
=====

* DbdOnto::Schema has Schema.org predicates with descriptions

0.0.7 (2013-10-20)
=====

* DbdOnto::Context and DbdOnto::Meta have URI and Comment properties

0.0.8 (2013-10-20)
=====

* Move the ontology documentation to ../data dir (and adapt the graphs)

0.0.11 (2013-10-23)
======

* no history for 0.0.9 and 0.0.10
* update to use Dbd 0.1.0

0.0.12 (2013-10-23)
======

* Add a Dbd ontology

0.0.13 (2014-02-09)
======

* a few more schema predicates to the used_predicate list
* fix the freeze feature on the Schema

0.0.14 (2014-03-17)
======

* 90+ additional schema predicates are added

0.0.15 (2014-04-17)
======

* Add DbdOnto::Rdf with rdf:type

0.0.16 (2014-04-24)
======

* Meta predicates also have meta:predicated_used



lib/dbd_onto/context.rb
module DbdOnto
  class Context < Base

    def initialize
      super
      self << meta_context
      self << context_resource
    end

  private

    def context_resource
      ::Dbd::Graph.new.from_CSV(context_resource_csv)
    end

    def context_resource_csv
<<EOS
"2013-09-04 20:06:29.398680722 UTC","49a95413-64b6-4f86-ad44-caa7b03931f6","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","03e72b0c-041f-4afd-aab4-520db49b1fb1","meta:defines_predicate","s","context:visibility"
"2013-09-04 20:06:29.398711938 UTC","1614f36e-5b64-43ef-9228-ef391dc2902a","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","03e72b0c-041f-4afd-aab4-520db49b1fb1","rdfs:label","s","Visibility"
"2013-09-04 20:06:29.398712938 UTC","1614f36e-5b64-43ef-9228-ef391dc2902b","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","03e72b0c-041f-4afd-aab4-520db49b1fb1","dbd:rdf_uri","s","https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md"
"2013-09-04 20:06:29.398713938 UTC","1614f36e-5b64-43ef-9228-ef391dc2902c","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","03e72b0c-041f-4afd-aab4-520db49b1fb1","rdfs:comment","s","E.g. public, private, professional, ..."
"2013-09-04 20:06:29.398728011 UTC","886738b0-fcf8-455e-a5a1-3f192b3cbbc1","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","735ae2a2-ed1f-405a-808a-07117cd6888c","meta:defines_predicate","s","context:encryption"
"2013-09-04 20:06:29.398743950 UTC","b85ea90d-01ea-4fc6-81b1-24cdabe9a968","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","735ae2a2-ed1f-405a-808a-07117cd6888c","rdfs:label","s","Encryption"
"2013-09-04 20:06:29.398744950 UTC","b85ea90d-01ea-4fc6-81b1-24cdabe9a969","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","735ae2a2-ed1f-405a-808a-07117cd6888c","dbd:rdf_uri","s","https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md"
"2013-09-04 20:06:29.398745950 UTC","b85ea90d-01ea-4fc6-81b1-24cdabe9a96a","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","735ae2a2-ed1f-405a-808a-07117cd6888c","rdfs:comment","s","E.g. clear, encrypted, AES-256, my_very_secret_key"
"2013-09-04 20:06:29.398758526 UTC","6d9b18f7-ed98-488e-8d67-bd830c51efbd","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","5b2c73e5-490c-47e0-a014-592d8665ba58","meta:defines_predicate","s","context:license"
"2013-09-04 20:06:29.398775364 UTC","cd7d3d4a-d9c4-46e5-ba12-bc41938e50d4","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","5b2c73e5-490c-47e0-a014-592d8665ba58","rdfs:label","s","License"
"2013-09-04 20:06:29.398776364 UTC","cd7d3d4a-d9c4-46e5-ba12-bc41938e50d5","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","5b2c73e5-490c-47e0-a014-592d8665ba58","dbd:rdf_uri","s","https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md"
"2013-09-04 20:06:29.398777364 UTC","cd7d3d4a-d9c4-46e5-ba12-bc41938e50d6","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","5b2c73e5-490c-47e0-a014-592d8665ba58","rdfs:comment","s","E.g. Copyright 2013 Peter Vandenabeele (http://vandenabeele.com), licensed under MIT"
"2013-09-04 20:06:29.411179884 UTC","a95a1e01-5701-41fa-ae57-5a93fc4e530e","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","ffc41d9c-ba51-4121-ae01-c2b04234b5f0","meta:defines_predicate","s","dc:source"
"2013-09-04 20:06:29.411232005 UTC","f063bd74-f767-47ed-a561-842eba382b79","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","ffc41d9c-ba51-4121-ae01-c2b04234b5f0","rdfs:label","s","Source"
"2013-09-04 20:06:29.411233005 UTC","f063bd74-f767-47ed-a561-842eba382b7a","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","ffc41d9c-ba51-4121-ae01-c2b04234b5f0","dbd:rdf_uri","s","https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md"
"2013-09-04 20:06:29.411234005 UTC","f063bd74-f767-47ed-a561-842eba382b7b","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","ffc41d9c-ba51-4121-ae01-c2b04234b5f0","rdfs:comment","s","E.g. Entered manually by @peter_v (also see Dublin Core)"
"2013-09-04 20:06:29.411249613 UTC","8ebc54af-bfd9-47ad-8e19-c04282720743","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","2b5008b0-e12d-4aa2-8bb1-f9c20c36f8ef","meta:defines_predicate","s","dc:creator"
"2013-09-04 20:06:29.411264605 UTC","6d726d47-31cb-43a9-9e6d-b44401149500","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","2b5008b0-e12d-4aa2-8bb1-f9c20c36f8ef","rdfs:label","s","Creator"
"2013-09-04 20:06:29.411265605 UTC","6d726d47-31cb-43a9-9e6d-b44401149501","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","2b5008b0-e12d-4aa2-8bb1-f9c20c36f8ef","dbd:rdf_uri","s","https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md"
"2013-09-04 20:06:29.411266605 UTC","6d726d47-31cb-43a9-9e6d-b44401149502","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","2b5008b0-e12d-4aa2-8bb1-f9c20c36f8ef","rdfs:comment","s","E.g. @peter_v (also see Dublin Core)"
"2013-09-04 20:06:29.414653441 UTC","90b35dd7-baa3-49b7-a224-d67d3ccadb7a","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","4aba46de-f35a-4754-913d-bdc4c926fc73","meta:defines_predicate","s","dcterms:created"
"2013-09-04 20:06:29.414677275 UTC","1ed030cf-260d-4380-bbbf-84f7c7e9d753","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","4aba46de-f35a-4754-913d-bdc4c926fc73","rdfs:label","s","Created"
"2013-09-04 20:06:29.414678275 UTC","1ed030cf-260d-4380-bbbf-84f7c7e9d753","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","4aba46de-f35a-4754-913d-bdc4c926fc73","dbd:rdf_uri","s","https://github.com/petervandenabeele/dbd_onto/blob/master/data/context.md"
"2013-09-04 20:06:29.414679275 UTC","1ed030cf-260d-4380-bbbf-84f7c7e9d753","36c0b50f-834e-45f5-a911-e3f2b47fe4b9","4aba46de-f35a-4754-913d-bdc4c926fc73","rdfs:comment","s","E.g. 2013-10-20 (also see Dublin Core Terms)"
EOS
    end
  end
end
